sql: TestDistributedMergeResumePreservesProgress failed

sql.TestDistributedMergeResumePreservesProgress [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/21057269?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/21057269?buildTab=artifacts#/) on master @ [e4700ffda4936fb8b5c749eff2f2ed2985739499](https://github.com/cockroachdb/cockroach/commits/e4700ffda4936fb8b5c749eff2f2ed2985739499):


```
=== RUN   TestDistributedMergeResumePreservesProgress
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/90c452cb2aa35b6c8949bd4a1d72e11e/logTestDistributedMergeResumePreservesProgress2392179474
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:183: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    test_server_shim.go:679: DRPC is enabled (via override by TestingGlobalDRPCOption)
    indexbackfiller_test.go:1466: [fast checkpoint pause during map] before resuming backfill, checkpointed spans: []
    indexbackfiller_test.go:1430: [fast checkpoint pause during map] before index backfill progress update, completed spans: [/Table/104/{1-2}]
    indexbackfiller_test.go:1436: [fast checkpoint pause during map] after distributed merge map phase, manifests count: 1
    indexbackfiller_test.go:1447: [fast checkpoint pause during map] after distributed merge iteration 1, manifests count: 12
    indexbackfiller_test.go:1447: [fast checkpoint pause during map] after distributed merge iteration 2, manifests count: 0
    indexbackfiller_test.go:1450: [fast checkpoint pause during map] final iteration 2 completed (direct KV ingest)
    indexbackfiller_test.go:1466: [slow checkpoint pause during map] before resuming backfill, checkpointed spans: []
    indexbackfiller_test.go:1430: [slow checkpoint pause during map] before index backfill progress update, completed spans: [/Table/104/{1-2}]
    indexbackfiller_test.go:1436: [slow checkpoint pause during map] after distributed merge map phase, manifests count: 1
    indexbackfiller_test.go:1447: [slow checkpoint pause during map] after distributed merge iteration 1, manifests count: 12
    indexbackfiller_test.go:1447: [slow checkpoint pause during map] after distributed merge iteration 2, manifests count: 0
    indexbackfiller_test.go:1450: [slow checkpoint pause during map] final iteration 2 completed (direct KV ingest)
    indexbackfiller_test.go:1436: [pause after map phase before merge] after distributed merge map phase, manifests count: 1
    indexbackfiller_test.go:1439: [pause after map phase before merge] map phase complete, blocking to allow test to pause job
    indexbackfiller_test.go:1441: [pause after map phase before merge] unblocked by test, continuing with merge iterations
    indexbackfiller_test.go:1436: [pause after map phase before merge] after distributed merge map phase, manifests count: 1
    indexbackfiller_test.go:1439: [pause after map phase before merge] map phase complete, blocking to allow test to pause job
    indexbackfiller_test.go:1441: [pause after map phase before merge] unblocked by test, continuing with merge iterations
    indexbackfiller_test.go:1447: [pause after map phase before merge] after distributed merge iteration 1, manifests count: 12
    indexbackfiller_test.go:1447: [pause after map phase before merge] after distributed merge iteration 2, manifests count: 0
    indexbackfiller_test.go:1450: [pause after map phase before merge] final iteration 2 completed (direct KV ingest)
    indexbackfiller_test.go:1436: [pause after iteration 1 with checkpoint] after distributed merge map phase, manifests count: 1
    indexbackfiller_test.go:1447: [pause after iteration 1 with checkpoint] after distributed merge iteration 1, manifests count: 12
    indexbackfiller_test.go:1457: [pause after iteration 1 with checkpoint] iteration 1 complete, blocking to allow test to pause job
    indexbackfiller_test.go:1459: [pause after iteration 1 with checkpoint] unblocked by test, continuing with remaining iterations
    indexbackfiller_test.go:1436: [pause after iteration 1 with checkpoint] after distributed merge map phase, manifests count: 1
    indexbackfiller_test.go:1709: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/90c452cb2aa35b6c8949bd4a1d72e11e/logTestDistributedMergeResumePreservesProgress2392179474
--- FAIL: TestDistributedMergeResumePreservesProgress (37.69s)
=== RUN   TestDistributedMergeResumePreservesProgress/pause_after_iteration_1_with_checkpoint
    indexbackfiller_test.go:1543: checkpoint interval set to 10ms, waitForCheckpoint=true
    indexbackfiller_test.go:1386: checkpoint contains 1 SST manifests
    indexbackfiller_test.go:1569: condition failed to evaluate within 30s: from indexbackfiller_test.go:1576: expected job to be succeeded, but found status: failed
--- FAIL: TestDistributedMergeResumePreservesProgress/pause_after_iteration_1_with_checkpoint (31.47s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDistributedMergeResumePreservesProgress.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-59388

--- a/pkg/sql/index_backfiller.go
+++ b/pkg/sql/index_backfiller.go
@@ -30,7 +30,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/admission/admissionpb"
 	"github.com/cockroachdb/cockroach/pkg/util/buildutil"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
-	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/errors"
 	gogotypes "github.com/gogo/protobuf/types"
@@ -525,9 +524,6 @@ func (ib *IndexBackfillPlanner) runDistributedMerge(
 				return err
 			}
 
-			// Clean up SST files from the previous iteration.
-			ib.cleanupPreviousIterationSSTs(ctx, job.ID(), progress.SSTStoragePrefixes, iteration)
-
 			// Call testing knob for final iteration (manifests will be nil/empty).
 			if fn := ib.execCfg.DistSQLSrv.TestingKnobs.AfterDistributedMergeIteration; fn != nil {
 				fn(ctx, iteration, progress.SSTManifests)
@@ -566,9 +562,6 @@ func (ib *IndexBackfillPlanner) runDistributedMerge(
 			return err
 		}
 
-		// Clean up SST files from the previous stage.
-		ib.cleanupPreviousIterationSSTs(ctx, job.ID(), progress.SSTStoragePrefixes, iteration)
-
 		// Call testing knob after updating progress for this iteration.
 		if fn := ib.execCfg.DistSQLSrv.TestingKnobs.AfterDistributedMergeIteration; fn != nil {
 			fn(ctx, iteration, progress.SSTManifests)
@@ -579,32 +572,4 @@ func (ib *IndexBackfillPlanner) runDistributedMerge(
 	}
 
 	return nil
-}
-
-// cleanupPreviousIterationSSTs removes SST files from the previous stage of the
-// distributed merge. This is a best-effort operation; errors are logged but
-// don't fail the job.
-func (ib *IndexBackfillPlanner) cleanupPreviousIterationSSTs(
-	ctx context.Context, jobID jobspb.JobID, storagePrefixes []string, completedIteration int,
-) {
-	if len(storagePrefixes) == 0 {
-		return
-	}
-
-	subdirectory := bulkutil.NewDistMergePaths(jobID).InputSubdir(completedIteration)
-
-	cleaner := bulkutil.NewBulkJobCleaner(
-		ib.execCfg.DistSQLSrv.ExternalStorageFromURI,
-		username.NodeUserName(),
-	)
-	defer func() {
-		if err := cleaner.Close(); err != nil {
-			log.Ops.Warningf(ctx, "error closing bulk job cleaner: %v", err)
-		}
-	}()
-
-	if err := cleaner.CleanupJobSubdirectory(ctx, jobID, storagePrefixes, subdirectory); err != nil {
-		log.Ops.Warningf(ctx, "failed to cleanup previous iteration SSTs (job %d, subdirectory %s): %v",
-			jobID, subdirectory, err)
-	}
 }

--- a/pkg/sql/indexbackfiller_test.go
+++ b/pkg/sql/indexbackfiller_test.go
@@ -1545,6 +1545,7 @@ func waitForCheckpointPersisted(
 	expectedSpans []roachpb.Span,
 	expectedPhase int32,
 ) {
+	t.Helper()
 	testutils.SucceedsWithin(t, func() error {
 		stmt := `SELECT payload FROM crdb_internal.system_jobs WHERE id = $1`
 		var payloadBytes []byte
@@ -1586,7 +1587,7 @@ func waitForCheckpointPersisted(
 			return errors.Errorf("waiting for phase %d, current: %d",
 				expectedPhase, bp.DistributedMergePhase)
 		}
-		t.Logf("checkpoint contains %d SST manifests", len(bp.SSTManifests))
+		t.Logf("checkpoint contains %d SST manifests, phase=%d", len(bp.SSTManifests), bp.DistributedMergePhase)
 		return nil
 	}, 5*time.Second)
 }

--- a/pkg/sql/schemachanger/scdeps/BUILD.bazel
+++ b/pkg/sql/schemachanger/scdeps/BUILD.bazel
@@ -11,6 +11,7 @@ go_library(
     importpath = "github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scdeps",
     visibility = ["//visibility:public"],
     deps = [
+        "//pkg/cloud",
         "//pkg/config/zonepb",
         "//pkg/jobs",
         "//pkg/jobs/jobspb",
@@ -22,6 +23,7 @@ go_library(
         "//pkg/server/telemetry",
         "//pkg/settings",
         "//pkg/settings/cluster",
+        "//pkg/sql/bulkutil",
         "//pkg/sql/catalog",
         "//pkg/sql/catalog/catalogkeys",
         "//pkg/sql/catalog/catpb",
@@ -48,6 +50,7 @@ go_library(
         "//pkg/sql/stats",
         "//pkg/sql/types",
         "//pkg/util/admission/admissionpb",
+        "//pkg/util/log",
         "//pkg/util/syncutil",
         "//pkg/util/timeutil",
         "//pkg/util/uuid",

--- a/pkg/sql/schemachanger/scdeps/run_deps.go
+++ b/pkg/sql/schemachanger/scdeps/run_deps.go
@@ -10,10 +10,13 @@ import (
 	"fmt"
 	"math"
 
+	"github.com/cockroachdb/cockroach/pkg/cloud"
 	"github.com/cockroachdb/cockroach/pkg/jobs"
 	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
 	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/security/username"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/sql/bulkutil"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descs"
 	"github.com/cockroachdb/cockroach/pkg/sql/isql"
@@ -22,6 +25,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scrun"
 	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
 	"github.com/cockroachdb/cockroach/pkg/sql/stats"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 )
@@ -48,44 +52,47 @@ func NewJobRunDependencies(
 	statements []string,
 	sessionData *sessiondata.SessionData,
 	kvTrace bool,
+	externalStorageFactory cloud.ExternalStorageFromURIFactory,
 ) scrun.JobRunDependencies {
 	return &jobExecutionDeps{
-		collectionFactory:     collectionFactory,
-		db:                    db,
-		backfiller:            backfiller,
-		spanSplitter:          spanSplitter,
-		merger:                merger,
-		rangeCounter:          rangeCounter,
-		eventLoggerFactory:    eventLoggerFactory,
-		jobRegistry:           jobRegistry,
-		job:                   job,
-		codec:                 codec,
-		settings:              settings,
-		testingKnobs:          testingKnobs,
-		statements:            statements,
-		indexValidator:        indexValidator,
-		commentUpdaterFactory: metadataUpdaterFactory,
-		sessionData:           sessionData,
-		kvTrace:               kvTrace,
-		statsRefresher:        statsRefresher,
-		tableStatsCache:       tableStatsCache,
+		collectionFactory:      collectionFactory,
+		db:                     db,
+		backfiller:             backfiller,
+		spanSplitter:           spanSplitter,
+		merger:                 merger,
+		rangeCounter:           rangeCounter,
+		eventLoggerFactory:     eventLoggerFactory,
+		jobRegistry:            jobRegistry,
+		job:                    job,
+		codec:                  codec,
+		settings:               settings,
+		testingKnobs:           testingKnobs,
+		statements:             statements,
+		indexValidator:         indexValidator,
+		commentUpdaterFactory:  metadataUpdaterFactory,
+		sessionData:            sessionData,
+		kvTrace:                kvTrace,
+		statsRefresher:         statsRefresher,
+		tableStatsCache:        tableStatsCache,
+		externalStorageFactory: externalStorageFactory,
 	}
 }
 
 type jobExecutionDeps struct {
-	collectionFactory     *descs.CollectionFactory
-	db                    descs.DB
-	statsRefresher        scexec.StatsRefresher
-	tableStatsCache       *stats.TableStatisticsCache
-	backfiller            scexec.Backfiller
-	spanSplitter          scexec.IndexSpanSplitter
-	merger                scexec.Merger
-	commentUpdaterFactory MetadataUpdaterFactory
-	rangeCounter          backfiller.RangeCounter
-	eventLoggerFactory    func(isql.Txn) scrun.EventLogger
-	jobRegistry           *jobs.Registry
-	job                   *jobs.Job
-	kvTrace               bool
+	collectionFactory      *descs.CollectionFactory
+	db                     descs.DB
+	statsRefresher         scexec.StatsRefresher
+	tableStatsCache        *stats.TableStatisticsCache
+	backfiller             scexec.Backfiller
+	spanSplitter           scexec.IndexSpanSplitter
+	merger                 scexec.Merger
+	commentUpdaterFactory  MetadataUpdaterFactory
+	rangeCounter           backfiller.RangeCounter
+	eventLoggerFactory     func(isql.Txn) scrun.EventLogger
+	jobRegistry            *jobs.Registry
+	job                    *jobs.Job
+	kvTrace                bool
+	externalStorageFactory cloud.ExternalStorageFromURIFactory
 
 	indexValidator scexec.Validator
 
@@ -112,6 +119,17 @@ func (d *jobExecutionDeps) ClusterSettings() *cluster.Settings {
 func (d *jobExecutionDeps) WithTxnInJob(ctx context.Context, fn scrun.JobTxnFunc) error {
 	var createdJobs []jobspb.JobID
 	var tableStatsToRefresh []descpb.ID
+	// Create cleaner if external storage factory is available.
+	var cleaner *bulkutil.BulkJobCleaner
+	if d.externalStorageFactory != nil {
+		cleaner = bulkutil.NewBulkJobCleaner(d.externalStorageFactory, username.NodeUserName())
+		defer func() {
+			if err := cleaner.Close(); err != nil {
+				log.Dev.Warningf(ctx, "error closing bulk job cleaner: %v", err)
+			}
+		}()
+	}
+
 	err := d.db.DescsTxn(ctx, func(
 		ctx context.Context, txn descs.Txn,
 	) error {
@@ -140,6 +158,7 @@ func (d *jobExecutionDeps) WithTxnInJob(ctx context.Context, fn scrun.JobTxnFunc
 				d.db,
 				pl.GetNewSchemaChange().BackfillProgress,
 				pl.GetNewSchemaChange().MergeProgress,
+				cleaner,
 			),
 			periodicProgressFlusher: backfiller.NewPeriodicProgressFlusherForIndexBackfill(d.settings),
 			statements:              d.statements,

--- a/pkg/sql/schemachanger/scexec/backfiller/BUILD.bazel
+++ b/pkg/sql/schemachanger/scexec/backfiller/BUILD.bazel
@@ -3,6 +3,7 @@ load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
 go_library(
     name = "backfiller",
     srcs = [
+        "cleanup.go",
         "periodic_progress_flusher.go",
         "progress.go",
         "tracker.go",
@@ -16,6 +17,7 @@ go_library(
         "//pkg/roachpb",
         "//pkg/settings/cluster",
         "//pkg/sql/backfill",
+        "//pkg/sql/bulkutil",
         "//pkg/sql/catalog/descpb",
         "//pkg/sql/isql",
         "//pkg/sql/schemachanger/scexec",
@@ -31,6 +33,7 @@ go_library(
 go_test(
     name = "backfiller_test",
     srcs = [
+        "cleanup_test.go",
         "periodic_progress_flusher_test.go",
         "progress_test.go",
         "tracker_test.go",

--- a/pkg/sql/schemachanger/scexec/backfiller/cleanup.go
+++ b/pkg/sql/schemachanger/scexec/backfiller/cleanup.go
@@ -1,0 +1,64 @@
+// Copyright 2026 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package backfiller
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
+	"github.com/cockroachdb/cockroach/pkg/sql/bulkutil"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/errors"
+)
+
+// jobSubdirectoryCleaner defines the interface for cleaning up SST subdirectories.
+// It is satisfied by *bulkutil.BulkJobCleaner.
+type jobSubdirectoryCleaner interface {
+	CleanupJobSubdirectory(
+		ctx context.Context, jobID jobspb.JobID, storagePrefixes []string, subdirectory string,
+	) error
+}
+
+// phaseTransition represents a detected phase change in the distributed merge
+// pipeline after checkpoint persistence.
+type phaseTransition struct {
+	TableID            descpb.ID
+	OldPhase           int32
+	NewPhase           int32
+	SSTStoragePrefixes []string
+}
+
+// phaseTransitionCleaner orchestrates SST cleanup for phase transitions.
+type phaseTransitionCleaner struct {
+	jobID   jobspb.JobID
+	cleaner jobSubdirectoryCleaner
+}
+
+// cleanupTransition performs SST cleanup for a single phase transition.
+func (c *phaseTransitionCleaner) cleanupTransition(
+	ctx context.Context, transition phaseTransition,
+) error {
+	log.Dev.Infof(ctx, "triggering SST cleanup for job %d, table %d after phase %d→%d transition",
+		c.jobID, transition.TableID, transition.OldPhase, transition.NewPhase)
+
+	// Determine which subdirectory to clean up based on completed phase:
+	// - newPhase=1: cleanup map/ (map phase output, input to iteration 1)
+	// - newPhase=2: cleanup merge/iter-1/ (iteration 1 output, input to iteration 2)
+	subdirectory := bulkutil.NewDistMergePaths(c.jobID).InputSubdir(int(transition.NewPhase))
+
+	if len(transition.SSTStoragePrefixes) == 0 {
+		log.Dev.Infof(ctx, "no storage prefixes found, skipping cleanup")
+		return nil
+	}
+
+	if err := c.cleaner.CleanupJobSubdirectory(ctx, c.jobID, transition.SSTStoragePrefixes, subdirectory); err != nil {
+		return errors.Wrapf(err, "cleaning up subdirectory %s", subdirectory)
+	}
+
+	log.Dev.Infof(ctx, "successfully cleaned up SSTs in %s", subdirectory)
+	return nil
+}

--- a/pkg/sql/schemachanger/scexec/backfiller/cleanup_test.go
+++ b/pkg/sql/schemachanger/scexec/backfiller/cleanup_test.go
@@ -1,0 +1,108 @@
+// Copyright 2026 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package backfiller
+
+import (
+	"context"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/stretchr/testify/require"
+)
+
+// TestPhaseTransitionCleaner verifies that phase transitions trigger cleanup
+// of the correct subdirectories based on the completed phase.
+func TestPhaseTransitionCleaner(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	const testJobID = jobspb.JobID(12345)
+	testStoragePrefixes := []string{"nodelocal://1/", "nodelocal://2/"}
+
+	tests := []struct {
+		name              string
+		transition        phaseTransition
+		expectedSubdir    string
+		shouldCallCleanup bool
+	}{
+		{
+			name: "phase 0→1 transition cleans map/",
+			transition: phaseTransition{
+				TableID:            descpb.ID(1),
+				OldPhase:           0,
+				NewPhase:           1,
+				SSTStoragePrefixes: testStoragePrefixes,
+			},
+			expectedSubdir:    "map/",
+			shouldCallCleanup: true,
+		},
+		{
+			name: "phase 1→2 transition cleans merge/iter-1/",
+			transition: phaseTransition{
+				TableID:            descpb.ID(1),
+				OldPhase:           1,
+				NewPhase:           2,
+				SSTStoragePrefixes: testStoragePrefixes,
+			},
+			expectedSubdir:    "merge/iter-1/",
+			shouldCallCleanup: true,
+		},
+		{
+			name: "empty storage prefixes skips cleanup",
+			transition: phaseTransition{
+				TableID:            descpb.ID(1),
+				OldPhase:           0,
+				NewPhase:           1,
+				SSTStoragePrefixes: nil,
+			},
+			expectedSubdir:    "",
+			shouldCallCleanup: false,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := context.Background()
+			mockCleaner := &mockBulkJobCleaner{}
+			cleaner := &phaseTransitionCleaner{
+				jobID:   testJobID,
+				cleaner: mockCleaner,
+			}
+
+			err := cleaner.cleanupTransition(ctx, tc.transition)
+			require.NoError(t, err)
+
+			if tc.shouldCallCleanup {
+				require.Equal(t, 1, mockCleaner.cleanupCallCount)
+				require.Equal(t, testJobID, mockCleaner.lastJobID)
+				require.Equal(t, testStoragePrefixes, mockCleaner.lastStoragePrefixes)
+				require.Equal(t, tc.expectedSubdir, mockCleaner.lastSubdirectory)
+			} else {
+				require.Equal(t, 0, mockCleaner.cleanupCallCount)
+			}
+		})
+	}
+}
+
+// mockBulkJobCleaner is a test double that tracks calls to CleanupJobSubdirectory.
+type mockBulkJobCleaner struct {
+	cleanupCallCount    int
+	lastJobID           jobspb.JobID
+	lastStoragePrefixes []string
+	lastSubdirectory    string
+}
+
+// CleanupJobSubdirectory records the call parameters for verification.
+func (m *mockBulkJobCleaner) CleanupJobSubdirectory(
+	_ context.Context, jobID jobspb.JobID, storagePrefixes []string, subdirectory string,
+) error {
+	m.cleanupCallCount++
+	m.lastJobID = jobID
+	m.lastStoragePrefixes = storagePrefixes
+	m.lastSubdirectory = subdirectory
+	return nil
+}

--- a/pkg/sql/schemachanger/scexec/backfiller/progress.go
+++ b/pkg/sql/schemachanger/scexec/backfiller/progress.go
@@ -157,8 +157,9 @@ func newBackfillProgress(codec keys.SQLCodec, bp scexec.BackfillProgress) *backf
 		EndKey: indexPrefix.PrefixEnd(),
 	}
 	return &backfillProgress{
-		BackfillProgress: bp,
-		totalSpan:        indexSpan,
+		BackfillProgress:                   bp,
+		totalSpan:                          indexSpan,
+		lastPersistedDistributedMergePhase: bp.DistributedMergePhase,
 	}
 }
 
@@ -181,6 +182,12 @@ type backfillProgress struct {
 	// totalSpan represents the complete span of the source index being
 	// backfilled.
 	totalSpan roachpb.Span
+
+	// lastPersistedDistributedMergePhase tracks the last successfully persisted
+	// DistributedMergePhase value. Used to detect when phase transitions have
+	// been committed to the database. This enables lazy cleanup of SST files from
+	// previous phases.
+	lastPersistedDistributedMergePhase int32
 }
 
 func (p backfillProgress) matches(bf scexec.Backfill) error {

--- a/pkg/sql/schemachanger/scexec/backfiller/tracker.go
+++ b/pkg/sql/schemachanger/scexec/backfiller/tracker.go
@@ -13,6 +13,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
 	"github.com/cockroachdb/cockroach/pkg/keys"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/bulkutil"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/isql"
 	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scexec"
@@ -43,7 +44,8 @@ type RangeCounter interface {
 // records the remaining spans of the source index to scan.
 type Tracker struct {
 	trackerConfig
-	codec keys.SQLCodec
+	codec   keys.SQLCodec
+	cleaner *phaseTransitionCleaner // Optional: nil if no cleanup needed
 
 	mu struct {
 		syncutil.Mutex
@@ -56,6 +58,10 @@ type Tracker struct {
 var _ scexec.BackfillerTracker = (*Tracker)(nil)
 
 // NewTracker constructs a new Tracker.
+//
+// If cleaner is non-nil, the tracker will perform SST cleanup when phase
+// transitions are detected. The caller is responsible for closing the cleaner.
+// If nil, cleanup is disabled.
 func NewTracker(
 	codec keys.SQLCodec,
 	counter RangeCounter,
@@ -63,13 +69,24 @@ func NewTracker(
 	db isql.DB,
 	jobBackfillProgress []jobspb.BackfillProgress,
 	jobMergeProgress []jobspb.MergeProgress,
+	cleaner *bulkutil.BulkJobCleaner,
 ) *Tracker {
-	return newTracker(
+	tr := newTracker(
 		codec,
 		newTrackerConfig(codec, counter, job, db),
 		convertFromJobBackfillProgress(codec, jobBackfillProgress),
 		convertFromJobMergeProgress(codec, jobMergeProgress),
 	)
+
+	// Configure cleanup if cleaner provided.
+	if cleaner != nil {
+		tr.cleaner = &phaseTransitionCleaner{
+			jobID:   job.ID(),
+			cleaner: cleaner,
+		}
+	}
+
+	return tr
 }
 
 func newTracker(
@@ -259,7 +276,58 @@ func (b *Tracker) FlushCheckpoint(ctx context.Context) error {
 		return false
 	})
 	log.Dev.Infof(ctx, "writing %d backfill checkpoints and %d merge checkpoints", len(bps), len(mps))
-	return b.writeCheckpoint(ctx, bps, mps)
+	if err := b.writeCheckpoint(ctx, bps, mps); err != nil {
+		return err
+	}
+
+	// After successful write, detect phase transitions and handle cleanup.
+	if b.cleaner != nil {
+		transitions := b.detectPhaseTransitions(ctx, bps)
+		for _, transition := range transitions {
+			if err := b.cleaner.cleanupTransition(ctx, transition); err != nil {
+				log.Ops.Warningf(ctx, "phase transition cleanup failed: %v", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// detectPhaseTransitions detects and records phase transitions that occurred
+// after checkpoint persistence. It returns the list of detected transitions.
+func (b *Tracker) detectPhaseTransitions(
+	ctx context.Context, bps []scexec.BackfillProgress,
+) []phaseTransition {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	var transitions []phaseTransition
+	for _, bp := range bps {
+		key := toBackfillKey(bp.Backfill)
+		p, exists := b.mu.backfillProgress[key]
+		if !exists {
+			continue
+		}
+
+		oldPhase := p.lastPersistedDistributedMergePhase
+		newPhase := bp.DistributedMergePhase
+
+		if oldPhase != newPhase {
+			log.Dev.Infof(ctx, "detected persisted distributed merge phase transition for table %d: phase %d → %d",
+				bp.Backfill.TableID, oldPhase, newPhase)
+
+			transitions = append(transitions, phaseTransition{
+				TableID:            bp.Backfill.TableID,
+				OldPhase:           oldPhase,
+				NewPhase:           newPhase,
+				SSTStoragePrefixes: bp.SSTStoragePrefixes,
+			})
+			// Update last persisted phase.
+			p.lastPersistedDistributedMergePhase = newPhase
+		}
+	}
+
+	return transitions
 }
 
 func (b *Tracker) getTableIndexBackfillProgress(bf scexec.Backfill) (backfillProgress, bool) {

--- a/pkg/sql/schemachanger/scjob/job.go
+++ b/pkg/sql/schemachanger/scjob/job.go
@@ -150,6 +150,7 @@ func (n *newSchemaChangeResumer) run(ctx context.Context, execCtxI interface{}) 
 		payload.Statement,
 		execCtx.SessionData(),
 		execCtx.ExtendedEvalContext().Tracing.KVTracingEnabled(),
+		execCfg.DistSQLSrv.ExternalStorageFromURI,
 	)
 
 	// If there are no descriptors left, then we can short circuit here.

